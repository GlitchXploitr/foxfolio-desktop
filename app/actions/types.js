// @flow

export type Action =
  | { type: 'ADD_SOURCE', source: Object }
  | { type: 'EDIT_SOURCE', source: Object, newSource: Object }
  | { type: 'REQUEST_TRANSACTIONS', source: Object }
  | { type: 'RECEIVE_TRANSACTIONS', exchange: string, trades: Object[], transfers: Object[] }
  | { type: 'FAILED_TRANSACTIONS', exchange: string, error: string }
  | { type: 'TICKER_UPDATE', ticker: Object }
  ;

export type Dispatch = (action: Action | ThunkAction | PromiseAction | Array<Action>) => any;
export type GetState = () => Object;
export type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;
export type PromiseAction = Promise<Action>;
