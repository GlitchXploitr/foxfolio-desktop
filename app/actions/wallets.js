// @flow
import type { ThunkAction } from './action.d';
import type { Wallet } from './wallet.d';

export function addWallet(wallet: Wallet): ThunkAction {
  return (dispatch) => {
    dispatch({
      type: 'ADD_WALLET',
      wallet,
    });
  };
}

export function editWallet(wallet: Wallet, newWallet: Wallet): ThunkAction {
  return dispatch => {
    dispatch({
      type: 'EDIT_WALLET',
      wallet,
      newWallet,
    });
  };
}
