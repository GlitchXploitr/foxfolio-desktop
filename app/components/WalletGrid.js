// @flow
import React, { Component } from 'react';
import { Button, Grid, Typography } from 'material-ui';
import { Add } from 'material-ui-icons';
import { withStyles } from 'material-ui/styles';
import type { Coinlist } from '../reducers/coinlist/types.d';
import WalletGridItem from './WalletGridItem';
import WalletDialog from './WalletDialog';
import type { Wallet } from '../reducers/wallets/types.d';

const styles = theme => ({
  button: {
    margin: theme.spacing.unit,
    color: theme.palette.background.default,
    position: 'absolute',
    right: 40,
    bottom: 40,
  },
});

type Props = {
  classes: any,
  coinlist: Coinlist,
  wallets: Wallet[],
  addWallet: (source: Wallet) => void,
  editWallet: (oldWallet: Wallet, newWallet: Wallet) => void,
  deleteWallet: (source: Wallet) => void
};

type State = {
  open: boolean,
  isNew: boolean,
  currentWallet: Wallet
};

class WalletGrid extends Component<Props, State> {
  state = {
    open: false,
    isNew: true,
    currentWallet: { currency: '', address: '', quantity: 0 },
  };

  addDialog = () => {
    this.setState({ open: true, isNew: true, currentWallet: { currency: '', address: '', quantity: 0 } });
  };

  editDialog = (wallet: Wallet) => {
    this.setState({ open: true, isNew: false, currentWallet: wallet });
  };

  closeDialog = () => {
    this.setState({ open: false });
  };

  saveDialog = (wallet) => {
    if (this.state.isNew) {
      this.props.addWallet(wallet);
    } else {
      this.props.editWallet(this.state.currentWallet, wallet);
    }
    this.closeDialog();
  };

  render() {
    const { wallets, classes, coinlist, deleteWallet } = this.props;

    return (
      <div className="container">
        <Typography type="headline">Wallets</Typography>
        <Grid container>
          {wallets
            .sort((a, b) => a.address.localeCompare(b.address))
            .map(wallet => (
              <Grid item key={wallet.currency + wallet.address} sm={12} md={6}>
                <WalletGridItem wallet={wallet} onEdit={this.editDialog} onDelete={deleteWallet} coinlist={coinlist}/>
              </Grid>
            ))}
        </Grid>
        <Button fab color="primary" aria-label="add" className={classes.button} onClick={this.addDialog}>
          <Add/>
        </Button>
        <WalletDialog
          open={this.state.open}
          wallet={this.state.currentWallet}
          close={this.closeDialog}
          save={this.saveDialog}
        />
      </div>
    );
  }
}

export default withStyles(styles)(WalletGrid);
