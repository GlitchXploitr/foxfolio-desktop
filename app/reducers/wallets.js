// @flow
import R from 'ramda';
import type { Action } from '../actions/types';

export type walletType = {
  currency: string,
  address: string
};

export default function wallets(state: walletType[] = [], action: Action) {
  switch (action.type) {
    case 'ADD_WALLET':
      return [...state, action.wallet];
    case 'EDIT_WALLET':
      return [...R.reject(R.equals(action.wallet), state), action.newWallet];
    default:
      return state;
  }
}
