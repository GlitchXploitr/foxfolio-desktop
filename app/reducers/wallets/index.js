// @flow
import R from 'ramda';
import type { Action } from '../../actions/action.d';
import type { Wallet } from './types.d';

export function wallets(state: Wallet[] = [], action: Action): Wallet[] {
  switch (action.type) {
    case 'ADD_WALLET':
      return [...state, action.wallet];
    case 'EDIT_WALLET':
      return [...R.reject(R.equals(action.wallet), state), action.newWallet];
    case 'DELETE_WALLET':
      return R.reject(R.equals(action.wallet), state);
    default:
      return state;
  }
}
