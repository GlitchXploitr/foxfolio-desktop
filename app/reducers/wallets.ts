import R from 'ramda';
import { Action } from '../actions/actions.types';
import { Wallet } from './wallets.types';

export function wallets(state: Wallet[] = [], action: Action): Wallet[] {
  switch (action.type) {
    case 'ADD_WALLET':
      return [...state, action.wallet];
    case 'EDIT_WALLET':
      return [...R.reject(R.equals(action.wallet), state), action.newWallet];
    case 'DELETE_WALLET':
      return R.reject(R.equals(action.wallet), state);
    default:
      return state;
  }
}
