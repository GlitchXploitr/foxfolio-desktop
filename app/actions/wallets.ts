import 'redux-thunk';
import { Dispatch } from 'redux';
import { Action, ThunkAction } from './actions.types';
import { Wallet } from 'reducers/wallets';
import { requestTickerUpdate } from './ticker';
import { GlobalState } from '../reducers';

export function addWallet(wallet: Wallet): ThunkAction {
  return (dispatch: Dispatch<GlobalState>) => {
    dispatch(requestTickerUpdate([wallet.currency]));
    dispatch({
      type: 'ADD_WALLET',
      wallet,
    });
  };
}

export function editWallet(wallet: Wallet, newWallet: Wallet): ThunkAction {
  return (dispatch: Dispatch<GlobalState>) => {
    dispatch(requestTickerUpdate([wallet.currency]));
    dispatch({
      type: 'EDIT_WALLET',
      wallet,
      newWallet,
    });
  };
}

export function deleteWallet(wallet: Wallet): Action {
  return {
    type: 'DELETE_WALLET',
    wallet,
  };
}
