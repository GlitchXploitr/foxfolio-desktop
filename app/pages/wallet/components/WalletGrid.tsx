import { Button, Grid, Typography, WithStyles } from 'material-ui';
import { Add } from 'material-ui-icons';
import { StyleRulesCallback, withStyles } from 'material-ui/styles';
import React, { Component } from 'react';
import { Coinlist } from '../../../reducers/coinlist';
import { Wallet } from '../../../reducers/wallets.types';
import WalletDialog from './WalletDialog';
import { WalletGridItem } from './WalletGridItem';

const styles: StyleRulesCallback = theme => ({
  button: {
    margin: theme.spacing.unit,
    color: theme.palette.background.default,
    position: 'absolute',
    right: 40,
    bottom: 40,
  },
});

interface Props {
  coinlist: Coinlist;
  wallets: Wallet[];
  addWallet: (source: Wallet) => any;
  editWallet: (oldWallet: Wallet, newWallet: Wallet) => any;
  deleteWallet: (source: Wallet) => any;
}

interface State {
  open: boolean;
  isNew: boolean;
  currentWallet: Wallet;
}

export const WalletGrid = withStyles(styles)(
  class extends Component<Props & WithStyles, State> {
    public state = {
      open: false,
      isNew: true,
      currentWallet: { currency: '', address: '', quantity: 0 },
    };

    public addDialog = () => {
      this.setState({
        open: true,
        isNew: true,
        currentWallet: { currency: '', address: '', quantity: 0 },
      });
    };

    public editDialog = (wallet: Wallet) => {
      this.setState({ open: true, isNew: false, currentWallet: wallet });
    };

    public closeDialog = () => {
      this.setState({ open: false });
    };

    public saveDialog = wallet => {
      if (this.state.isNew) {
        this.props.addWallet(wallet);
      } else {
        this.props.editWallet(this.state.currentWallet, wallet);
      }
      this.closeDialog();
    };

    public render() {
      const { wallets, classes, coinlist, deleteWallet } = this.props;

      return (
        <div className="container">
          <Typography type="headline">Wallets</Typography>
          <Grid container>
            {wallets.sort((a, b) => a.address.localeCompare(b.address)).map(wallet => (
              <Grid item key={wallet.currency + wallet.address} sm={12} md={6}>
                <WalletGridItem
                  wallet={wallet}
                  onEdit={this.editDialog}
                  onDelete={deleteWallet}
                  coinlist={coinlist}
                />
              </Grid>
            ))}
          </Grid>
          <Button
            fab
            color="primary"
            aria-label="add"
            className={classes.button}
            onClick={this.addDialog}
          >
            <Add />
          </Button>
          <WalletDialog
            open={this.state.open}
            wallet={this.state.currentWallet}
            close={this.closeDialog}
            save={this.saveDialog}
            coinlist={coinlist}
          />
        </div>
      );
    }
  }
);
