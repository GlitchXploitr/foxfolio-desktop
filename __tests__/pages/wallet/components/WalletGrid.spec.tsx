import { createShallow } from 'material-ui/test-utils';
import * as React from 'react';
import { WalletGrid } from '../../../../app/pages/wallet/components/WalletGrid';
import { Wallet } from '../../../../app/reducers/wallets.types';

let shallow;

beforeAll(() => {
  shallow = createShallow();
});

let wallets: Wallet[];

beforeEach(() => {
  wallets = [{ address: '', currency: 'BTC', quantity: 1 }];
});

test('Render wallet grid', () => {
  const coinlist = {};
  const addWallet = jest.fn();
  const editWallet = jest.fn();
  const deleteWallet = jest.fn();

  const wrapper = shallow(
    <WalletGrid
      coinlist={coinlist}
      wallets={wallets}
      addWallet={addWallet}
      editWallet={editWallet}
      deleteWallet={deleteWallet}
    />
  ).dive();
  expect(wrapper).toMatchSnapshot();
});
