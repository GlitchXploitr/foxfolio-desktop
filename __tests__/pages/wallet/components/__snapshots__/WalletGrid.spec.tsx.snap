// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render wallet grid 1`] = `
<div
  className="container"
>
  <WithStyles(Typography)
    variant="headline"
  >
    Wallets
  </WithStyles(Typography)>
  <WithStyles(GridWrapper)
    container={true}
    spacing={16}
  >
    <WithStyles(GridWrapper)
      item={true}
      key="BTC"
      lg={6}
      md={8}
      sm={12}
    >
      <WithStyles(Component)
        coinlist={Object {}}
        onDelete={[Function]}
        onEdit={[Function]}
        wallet={
          Object {
            "address": "",
            "currency": "BTC",
            "quantity": 1,
          }
        }
      />
    </WithStyles(GridWrapper)>
  </WithStyles(GridWrapper)>
  <WithStyles(Button)
    aria-label="add"
    className="Component-button-1"
    color="primary"
    onClick={[Function]}
    variant="fab"
  >
    <pure(Add) />
  </WithStyles(Button)>
  <WalletDialog
    close={[Function]}
    coinlist={Object {}}
    open={false}
    save={[Function]}
    wallet={
      Object {
        "address": "",
        "currency": "",
        "quantity": 0,
      }
    }
  />
</div>
`;
